### PR TITLE
[q4-working-branch] Q4 network plus

### PR DESCRIPTION
**Pull request by Steve-White-UK**
## Description

This PR is the latest stable version incorporating months of testing and Bartosz latest stable fixes to bases system,.

## Backward Compatibility

- [  This is not a breaking change
- [ X] This is a breaking change

## Review Suggestion

Develop with this

## Status
Not to be merged to master without strict review
### Implementation

- [ ] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libNetwork/RumorManager.cpp [hunk 1]
```diff
@@ -713,8 +716,12 @@ void RumorManager::AppendKeyAndSignature(RawBytes& result,
   tmp2.push_back((unsigned char)(CHAIN_ID & 0xFF));
   tmp2.insert(tmp2.end(), messageToSig.begin(), messageToSig.end());
 
-  Signature sig = P2PComm::GetInstance().SignMessage(tmp2);
-  sig.Serialize(tmp, PUB_KEY_SIZE);
+  auto sig = zil::p2p::GetInstance().SignMessage(tmp2);
```
Comment (Steve-White-UK) [inline src/libNetwork/RumorManager.cpp:719]: We need to investigate why this is here

File: primary_account.json
Comment (its-saeed) [inline primary_account.json:1]: Is it okay to have this file in the project's root directory?

**Review by its-saeed — APPROVED** [review]

File: src/libDirectoryService/DSBlockPreProcessing.cpp [hunk 2]
```diff
@@ -303,12 +304,12 @@ void DirectoryService::InjectPoWForDSNode(
...
     // Remove this node from blacklist if it exists
     Peer& p = rit->second;
-    Blacklist::GetInstance().Remove(p.GetIpAddress());
+    Blacklist::GetInstance().Remove({p.GetIpAddress(),p.GetListenPortHost(),p.GetNodeIndentifier()});
 
     ++counter;
   }
 
-  LOG_GENERAL(INFO, "Num PoWs after injection = " << sortedPoWSolns.size());
+  LOG_GENERAL(INFO, "### Num PoWs after injection = " << sortedPoWSolns.size());
```
Comment (its-saeed) [inline src/libDirectoryService/DSBlockPreProcessing.cpp:312]: Is it intentional?

File: src/libNetwork/Blacklist.h [hunk 3]
```diff
@@ -34,6 +34,30 @@ struct hash<uint128_t> {
 };
 }  // namespace std
 
+struct NodeKey {
+        uint128_t ip;
```
Comment (its-saeed) [inline src/libNetwork/Blacklist.h:38]: Seems indents are not correct in this struct

File: src/libNetwork/DataSender.cpp [hunk 4]
```diff
@@ -49,15 +50,15 @@ void SendDataToLookupNodesDefault(const VectorOfNode& lookups,
...
-    Peer tmp(resolved_ip, node.second.GetListenPortHost());
+    Blacklist::GetInstance().Whitelist({resolved_ip,node.second.GetListenPortHost(),node.second.GetNodeIndentifier()});  // exclude this lookup ip from blacklisting
+    Peer tmp(resolved_ip, node.second.GetListenPortHost(),node.second.GetNodeIndentifier());
     LOG_GENERAL(INFO, "Sending to lookup " << tmp);
 
     allLookupNodes.emplace_back(tmp);
   }
 
-  P2PComm::GetInstance().SendBroadcastMessage(allLookupNodes, message);
+  // XXX P2PComm::GetInstance().SendBroadcastMessage(allLookupNodes, message);
```
Comment (its-saeed) [inline src/libNetwork/DataSender.cpp:60]: shouldn't be removed?

**Review by Steve-White-UK — COMMENTED** [review]
LGTM

File: src/libNetwork/RumorManager.cpp [hunk 5]
```diff
@@ -318,8 +317,8 @@ bool RumorManager::AddForeignRumor(const RumorManager::RawBytes& message) {
                          SIGNATURE_RESPONSE_SIZE,
                      message.end());
 
-  if (!P2PComm::GetInstance().VerifyMessage(raw_message, toVerify,
-                                            senderPubKey)) {
+  // XXX
```
Comment (its-saeed) [inline src/libNetwork/RumorManager.cpp:320]: what does this mean?

File: src/libNetwork/RumorManager.cpp [hunk 6]
```diff
@@ -713,8 +716,12 @@ void RumorManager::AppendKeyAndSignature(RawBytes& result,
   tmp2.push_back((unsigned char)(CHAIN_ID & 0xFF));
   tmp2.insert(tmp2.end(), messageToSig.begin(), messageToSig.end());
 
-  Signature sig = P2PComm::GetInstance().SignMessage(tmp2);
-  sig.Serialize(tmp, PUB_KEY_SIZE);
+  auto sig = zil::p2p::GetInstance().SignMessage(tmp2);
+  if (!sig) {
+    // XXX log fatal?
```
Comment (its-saeed) [inline src/libNetwork/RumorManager.cpp:721]: ?

**Pull request merged**